src/NuGet.Clients/PackageManagement.UI/Services/NuGetLockService.cs

```diff
@@ -18,126 +17,49 @@ public sealed class NuGetLockService : INuGetLockService, IDisposable
     {
         private readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
 
-        public bool IsLockHeld => _semaphore.CurrentCount == 0;
+        private readonly AsyncLocal<int> _lockCount = new AsyncLocal<int>();
 
-        public IDisposable AcquireLock()
-        {
-            _semaphore.Wait();
-            return new SemaphoreLockReleaser(_semaphore);
-        }
-
-        public IAsyncLockAwaitable AcquireLockAsync(CancellationToken token)
-        {
-            return new SemaphoreLockAwaiter(_semaphore, token);
-        }
-
-        public void Dispose()
-        {
-            _semaphore.Dispose();
-        }
+        public bool IsLockHeld => _semaphore.CurrentCount == 0;
 
-        private class SemaphoreLockReleaser : IDisposable
+        public async Task<T> ExecuteNuGetOperationAsync<T>(Func<Task<T>> action, CancellationToken token)
         {
-            private readonly SemaphoreSlim _semaphore;
-            private bool _isDisposed = false;
-
-            public SemaphoreLockReleaser(SemaphoreSlim semaphore)
+            if (_lockCount.Value == 0)
             {
-                _semaphore = semaphore;
-            }
-
-            public void Dispose()
-            {
-                Dispose(true);
-                GC.SuppressFinalize(this);
-            }
-
-            private void Dispose(bool _)
-            {
-                if (_isDisposed)
-                {
-                    return;
-                }
-
+                _lockCount.Value++;
```

Comment:
It looks like you're just using the integer as a boolean. And the original semaphore acquirer will release the semaphore regardless of whether the nested users are done or not. That might be OK for you. But you might consider incrementing and decrementing this counter for every caller, and only wait for the semaphore when bumping to 1, and releasing when bumping to 0. That way, if nested work outlives its parent, it still holds the semaphore till it's done.
But again, that might be complexity that you don't need. So feel free to Won't Fix.